### DISCUSSION TITLE
Metadata-only operations

### DISCUSSION THREAD

**Discussion by olix0r** [kube-rs/kube#1023] [General]
I recently noticed that [client-go supports operations that only interact with resource metadata]( 
https://github.com/kubernetes/client-go/blob/be20b2bd91d8997dd6e42647be047973acb8cdee/metadata/metadata.go#L180-L306) via the [`PartialObjectMetadata` type](https://github.com/kubernetes/apimachinery/blob/7fb78ee962897d9de6bac4a8f0f1346eb1480ac4/pkg/apis/meta/v1/types.go#L1425-L1434). This can allow controllers to substantially reduce I/O and memory requirements when they don't actually need an object's spec or status.

I haven't found any analogs in the kube-rs ecosystem. Do they exist? If not, do you have any thoughts about where/how this should be implemented?

**Comment by clux** [comment]
It seems to be available on [Kubernetes > 1.14](https://pkg.go.dev/k8s.io/client-go/metadata#Client), although it's not listed as verbs in the openapi spec so it's gone under the radar. We should almost certainly add support for this.

The request there sets a header for a protobuf specific request, but if it works outside protobuf, then we could probably expose getters/setters on `Api` that treat it like other subresources (since that feels pretty similar here):

- `async fn get_metadata(name: &str) -> Result<ObjectMeta>`
- `async fn list_metadata(lp: &ListParams) -> Result<ObjectList<ObjectMeta>>` ? not sure if the return type is wrapped in the native list type
- `async fn patch_metadata(name: &str, pp: &PatchParams, patch: &Patch<P>) -> Result<ObjectMeta>`
- `async fn watch_metadata(lp: &ListParams, version: &str) -> Result<impl Stream<Item = Result<WatchEvent<ObjectMeta>>>>` maybe?  not sure how the final wrap would look like coming out of the api

If it doesn't work outside protobuf, then it's possible we could still introduce these as a first go for protobuf support by sketching out `ObjectMeta` protobufs correctly first. The types for it would exist in k8s-pb.

WDYT?

**Reply by olix0r** [reply]
> The request there sets a header for a protobuf specific request

True, but note that [`Accept` includes](https://github.com/kubernetes/client-go/blob/be20b2bd91d8997dd6e42647be047973acb8cdee/metadata/metadata.go#L186):

* `application/vnd.kubernetes.protobuf;as=PartialObjectMetadata;g=meta.k8s.io;v=v1`
* `application/json;as=PartialObjectMetadata;g=meta.k8s.io;v=v1`
* `application/json`

So, I'm hopeful we should be able to decouple this from protobuf support.

**Reply by olix0r** [reply]
Tracked in https://github.com/kube-rs/kube/issues/1027
